contracts/external/dao-migrator/src/contract.rs

```diff
@@ -0,0 +1,402 @@
+            {
+                contract_info.code_id
+            } else {
+                // Return false if we don't get contract info, means something went wrong.
+                return Err(ContractError::NoContractInfo {
+                    address: cw20_staked_addr.into(),
+                });
+            };
+
+            // If module is not DAO DAO module
+            if c20_staked_code_id != staking_pair.v1_code_id {
+                return Err(ContractError::CantMigrateModule {
+                    code_id: c20_staked_code_id,
+                });
+            }
+
+            msgs.push(
+                WasmMsg::Migrate {
+                    contract_addr: cw20_staked_addr.to_string(),
+                    new_code_id: staking_pair.v2_code_id,
+                    msg: to_binary(&staking_pair.migrate_msg).unwrap(),
+                }
+                .into(),
+            );
+        }
+    } else {
+        return Err(ContractError::VotingModuleNotFound);
+    }
+
+    // -----------------------
+    // verify proposal modules
+    // -----------------------
+    // We take all the proposal modules of the DAO.
+    let proposal_modules: Vec<ProposalModule> = deps.querier.query_wasm_smart(
+        info.sender.clone(),
+        &dao_core::msg::QueryMsg::ProposalModules {
+            start_after: None,
+            limit: None,
+        },
+    )?;
+
+    let success_proposal = proposal_modules.into_iter().all(|module| {
+        // Make sure that we don't try to migrate the migrator ...
+        if env.contract.address == module.address {
+            return true;
+        }
+        // Get the code id of the module
+        let proposal_code_id = if let Ok(contract_info) = deps
+            .querier
+            .query_wasm_contract_info(module.address.clone())
+        {
+            contract_info.code_id
+        } else {
+            // Return false if we don't get contract info, means something went wrong.
+            proposal_error = Some(ContractError::NoContractInfo {
+                address: module.address.into(),
+            });
+            return false;
+        };
+
+        // check if Code id is valid DAO DAO code id
+        if proposal_code_id == proposal_pair.v1_code_id {
+            msgs.push(
+                WasmMsg::Migrate {
+                    contract_addr: module.address.to_string(),
+                    new_code_id: proposal_pair.v2_code_id,
+                    msg: to_binary(&proposal_pair.migrate_msg).unwrap(),
+                }
+                .into(),
+            );
+            modules_addrs.proposals.push(module.address.to_string());
+        } else {
+            // Return false because we couldn't find the code id on our list.
+            proposal_error = Some(ContractError::CantMigrateModule {
+                code_id: proposal_code_id,
+            });
+            return false;
+        }
+
+        true
+    });
+
+    if !success_proposal {
+        Err(proposal_error.unwrap())
```

Comment:
So I was actually looking into that example, and if I have some simple map like in there, I have no problems with it, but for some reason I didn't manage to do it with a more complex map and custom errors (ContractError and not StdError).

Also I think I can't do a map here, because I need to ignore one of the entries of the vec, `if env.contract.address == module.address` we need to ignore the migrator module here.
Plus, I don't actually care about the vec after, just want to loop over it.

Reply:
> I need to ignore one of the entries of the vec, if env.contract.address == module.address

`iter().filter(|m| m.address != env.contract.address).map(/* ... */)`

my particular concern here is the interplay between `success_proposal` and `proposal_error` that is not expressed in the type system (and thus the need for the unwrap). if we, in the future, modify this contract and forget to correctly handle that interplay, that's a great way to have a bug.

i _think_ that using the collect approach we can move the management of that interplay to the rust standard library which is a much nicer place for it.

> for some reason I didn't manage to do it with a more complex map and custom errors (ContractError and not StdError)

this is definitely tricky. could you DM me the error that you get? i can help you fix this. I run into this often.

Reply:
Ha, some how I managed to do it on the first try, no idea why I was struggling with it at the first place, pretty straight forward honestly.

I do want to raise 2 questions about it:

1. I didn't do the filter, because I don't care about the vec afterward, so instead I just ignore our module, I just Ok(()) if it is the migrator module, that way, we only do 1 loop instead of 2.
2. from the `.map()` docs: `If you're doing some sort of looping for a side effect, it's considered more idiomatic to use for than map()`, how important is it in this case? seems like this sentence is talking exactly about my case here.

Reply:
Should have ran clippy before commenting (future note), clippy complained about it, and suggested to use `.try_for_each()` which make sense.